api/pkg/s3/objectput.go

```diff
@@ -15,92 +15,188 @@
+	var err error
+	if !isValidObjectName(objectKey) {
+		response.WriteError(http.StatusBadRequest, s3error.ErrInvalidObjectName)
+		return
+	}
+
+	// get size
+	size, err := getSize(request, response)
 	if err != nil {
-		log.Errorf("get content length failed, err: %v\n", err)
-		response.WriteError(http.StatusInternalServerError, InvalidContentLength.Error())
 		return
 	}
-	backendName := request.HeaderParameter("x-amz-storage-class")
-	log.Infof("object.size is %v, objectKey is %s, backend name is%s\n", size, objectKey, backendName)
-
-	// Currently, only support tier1 as default
-	tier := int32(utils.Tier1)
-
-	object := s3.Object{}
-	object.BucketName = bucketName
-	object.ObjectKey = objectKey
-	object.Size = size
-	object.IsDeleteMarker = ""
-	object.InitFlag = ""
-	object.LastModified = time.Now().Unix()
-	object.Tier = tier
-	// standard as default
-	object.StorageClass = constants.StorageClassOpenSDSStandard
-
-	md := map[string]string{common.REST_KEY_OPERATION: common.REST_VAL_UPLOAD}
-	ctx := common.InitCtxWithVal(request, md)
-	actx := request.Attribute(c.KContext).(*c.Context)
-	object.UserId = actx.UserId
-	object.TenantId = actx.TenantId
-
-	var client datastore.DataStoreAdapter
-	if backendName != "" {
-		object.Backend = backendName
-		client = getBackendByName(ctx, s, backendName)
+	if size == -1 {
+		response.WriteError(http.StatusLengthRequired, s3error.ErrMissingContentLength)
+		return
+	}
+
+	// maximum Upload size for objects in a single operation
+	if isMaxObjectSize(size) {
+		response.WriteError(http.StatusBadRequest, s3error.ErrEntityTooLarge)
+		return
+	}
+
+	// Save metadata.
+	metadata := extractMetadataFromHeader(request)
+	// Get Content-Md5 sent by client and verify if valid
+	if _, ok := request.Request.Header["Content-Md5"]; !ok {
+		metadata["md5Sum"] = ""
 	} else {
-		bucket, _ := s.s3Client.GetBucket(ctx, &s3.Bucket{Name: bucketName})
-		object.Backend = bucket.Backend
-		client = getBackendClient(ctx, s, bucketName)
+		if len(request.Request.Header.Get("Content-Md5")) == 0 {
+			log.Infoln("Content Md5 is null!")
+			response.WriteError(http.StatusBadRequest, s3error.ErrInvalidDigest)
+			return
+		}
+		md5Bytes, err := checkValidMD5(request.Request.Header.Get("Content-Md5"))
+		if err != nil {
+			log.Infoln("Content Md5 is invalid!")
+			response.WriteError(http.StatusBadRequest, s3error.ErrInvalidDigest)
+			return
+		} else {
+			metadata["md5Sum"] = hex.EncodeToString(md5Bytes)
+		}
 	}
 
-	if client == nil {
-		response.WriteError(http.StatusInternalServerError, NoSuchBackend.Error())
+	acl, err := getAclFromHeader(request)
+	if err != nil {
+		WriteErrorResponse(request, response, err)
 		return
 	}
-	log.Infof("enter the PUT method")
-	s3err := client.PUT(request.Request.Body, &object, ctx)
-	log.Infof("LastModified is %v\n", object.LastModified)
-	if s3err != NoError {
-		response.WriteError(http.StatusInternalServerError, s3err.Error())
+
+	// check if specific bucket exist
```

Comment:
line 103 to 118, check if backend exists and check if bucket exists, will be common for many bucket operations, please extract a method (refactor)

Reply:
@rhsakarpos ,  check if backend exists is only required when put object and put backend,  I think refactor is not necessary for now.  What do you think about this?

Reply:
i saw the same code repeated below, please refactor the 2 methods for backend exist check and bucket exist check

Reply:
@rhsakarpos , Thanks, I will take your advice, I add your suggestion to issue, and can I do it at next PR?